[WebDriver][BiDi] Implement the browser.removeUserContext command

yury-s opened the pull request:
#### 75d2c8aeeeb139793f6e1a4d873d4f4e038d06e8
<pre>
[WebDriver][BiDi] Implement the browser.removeUserContext command
<a href="https://bugs.webkit.org/show_bug.cgi?id=288108">https://bugs.webkit.org/show_bug.cgi?id=288108</a>

Reviewed by BJ Burg.

Close existing pages when user context is being removed and wait
for them to actually close before responding to the command.
If new WebPageProxy is created while its user context is being
removed, the page is closed immediately.

* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
(WebKit::BidiBrowserAgent::BidiUserContextDeletionRecord::BidiUserContextDeletionRecord):
(WebKit::BidiBrowserAgent::didCreatePage):
(WebKit::BidiBrowserAgent::willClosePage):
(WebKit::BidiBrowserAgent::close):
(WebKit::BidiBrowserAgent::createUserContext):
(WebKit::BidiBrowserAgent::getUserContexts):
(WebKit::BidiBrowserAgent::removeUserContext):
* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.h:
* Source/WebKit/UIProcess/Automation/BidiUserContext.cpp:
(WebKit::BidiUserContext::allPages const):
* Source/WebKit/UIProcess/Automation/BidiUserContext.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::didCreatePage):
(WebKit::WebAutomationSession::willClosePage):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_pageForTesting):

Canonical link: <a href="https://commits.webkit.org/294938@main">https://commits.webkit.org/294938@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c25eaac968586ef5472abe1679648f606e252465

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101238 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20900 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11203 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106386 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51865 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21209 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29394 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77098 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34141 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104245 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16354 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91440 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57445 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16170 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51213 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86055 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9526 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108742 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28366 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20874 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86077 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28728 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87648 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85627 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22231 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30359 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8066 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22465 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28296 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33567 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28108 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31428 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29666 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
```diff
@@ -33,6 +33,7 @@
 #include "APIOpenPanelParameters.h"
 #include "APIString.h"
 #include "AutomationProtocolObjects.h"
+#include "BidiBrowserAgent.h"
```
Belongs inside ENABLE(WEBDRIVER_BIDI) below.

burg commented on Source/WebKit/UIProcess/Automation/BidiUserContext.h:
```diff
@@ -56,6 +57,8 @@ class BidiUserContext {
     WebsiteDataStore& dataStore() const { return m_dataStore.get(); }
     WebProcessPool& processPool() const { return m_processPool.get(); }
 
+    Vector<Ref<WebPageProxy>> pages() const;
```
Naming Nit: this seems *too* short to me. maybe `associatedPages()` or `relatedPages()` or `allPages()`?

yury-s replied:
Renamed to `allPages()`.

burg commented on Source/WebKit/UIProcess/Automation/BidiUserContext.cpp:
```diff
@@ -50,6 +52,18 @@ BidiUserContext::BidiUserContext(WebsiteDataStore& dataStore, WebProcessPool& pr
+    Vector<Ref<WebPageProxy>> pages;
+    for (Ref process : m_processPool->processes()) {
+        for (Ref page : process->pages()) {
+            if (page->websiteDataStore() == m_dataStore.get())
```
This is a suggestion, but it would be nice to assert here (or in `createUseContext()`?) that the websiteDataStore for a UserContext is ephemeral.

yury-s replied:
Added an assert to `createUserContext`. I think this class can be used to represent the default context too.

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -110,18 +164,23 @@ Inspector::CommandResult<Ref<JSON::ArrayOf<UserContextInfo>>> BidiBrowserAgent::
 }
 
-Inspector::CommandResult<void> BidiBrowserAgent::removeUserContext(const String& userContext)
+void BidiBrowserAgent::removeUserContext(const String& userContext, Inspector::CommandCallback<void>&& callback)
```
Nit: Inspector namespace not necessary in this file.

burg replied:
Naming: `userContextID`

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -110,18 +164,23 @@ Inspector::CommandResult<Ref<JSON::ArrayOf<UserContextInfo>>> BidiBrowserAgent::
-    // TODO: track and close all pages that belong to this user context.
-    m_userContexts.remove(it);
-    return { };
+    auto context = m_userContexts.take(it);
```
Naming: I would prefer this one be `userContext` since it's strongly typed

yury-s replied:
Renamed.

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -110,18 +164,23 @@ Inspector::CommandResult<Ref<JSON::ArrayOf<UserContextInfo>>> BidiBrowserAgent::
-    m_userContexts.remove(it);
-    return { };
+    auto context = m_userContexts.take(it);
+    size_t numberOfPages = context->pages().size();
```
Naming: `pageCount`

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.h:
```diff
@@ -51,18 +52,23 @@ class BidiBrowserAgent final : public Inspector::BidiBrowserBackendDispatcherHan
     WeakPtr<WebAutomationSession> m_session;
     Ref<Inspector::BidiBrowserBackendDispatcher> m_browserDomainDispatcher;
     HashMap<String, std::unique_ptr<BidiUserContext>> m_userContexts;
+    HashMap<String, std::unique_ptr<UserContextDeletion>> m_userContextDeletions;
```
Naming: I know it's more wordy, but `BidiUserContextDeletionRecord` and `m_userContextsPendingDeletion` will make it way more obvious how this works. 🙏🏻

If this becomes a standalone class, I'd add `Bidi` prefix to match `BidiUserContext`.

yury-s replied:
Renamed both.

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -61,6 +64,46 @@ const String& defaultUserContextID()
+public:
+    UserContextDeletion(std::unique_ptr<BidiUserContext>&& context, size_t numberOfPages, Inspector::CommandCallback<void>&& callback)
+        : m_userContext(WTFMove(context))
+        , m_numberOfPages(numberOfPages)
```
Naming: `m_pageCount`

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -61,6 +64,46 @@ const String& defaultUserContextID()
 
 } // namespace
```
Nit: `} // namespace WebKit` or remove the comment

yury-s replied:
This one actually closes an anonymous namespace inside WebKit, so I believe it should stay `// namespace` as in other places in the codebase.

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -61,6 +64,46 @@ const String& defaultUserContextID()
+    WTF_MAKE_NONCOPYABLE(UserContextDeletion);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    UserContextDeletion(std::unique_ptr<BidiUserContext>&& context, size_t numberOfPages, Inspector::CommandCallback<void>&& callback)
```
Naming: `userContext`

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -61,6 +64,46 @@ const String& defaultUserContextID()
+        , m_numberOfPages(numberOfPages)
+        , m_callback(WTFMove(callback)) { }
+
+    void didDestroyPage(const WebPageProxy& page)
```
Two pieces of feedback:

1. I'd prefer that we use this class as a simple "record" of pending deletion and put the bookkeeping logic into BidiBrowserAgent private methods. It will be easier to follow the entire lifecycle rather than splitting up willCreatePage and didDestroyPage.

2. It seems kind of weird to keep a count of open pages and check against an old value, shouldn't we know when new pages are created? If a page is created and there is a pending BidiUserContextDeletionRecord for the associated user context, we should append the doomed page to a page deletion worklist. Dequeue and close each page until there are no more, then we can call `callback` to finish the overall `removeUserContext` command. It looks like you'd need to add another hook from `WebPageProxy::initializeWebPage` similar to the instrumentation for `willClosePage`.

yury-s replied:
Done both. I place `didCreatePage` hook in `WebPageProxy` constructor though, otherwise it may happen that there is a page which doesn't have a web process yet, but we already report its user context as removed.

the-chenergy commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -110,18 +156,23 @@ Inspector::CommandResult<Ref<JSON::ArrayOf<UserContextInfo>>> BidiBrowserAgent::
-    m_userContexts.remove(it);
-    return { };
+    auto userContext = m_userContexts.take(it);
+    size_t numberOfPages = userContext->allPages().size();
```
```suggestion
    size_t pageCount = userContext->allPages().size();
```

yury-s replied:
Done.

yury-s committed:
[WebDriver][BiDi] Implement the browser.removeUserContext command
https://bugs.webkit.org/show_bug.cgi?id=288108

Reviewed by BJ Burg.

Close existing pages when user context is being removed and wait
for them to actually close before responding to the command.
If new WebPageProxy is created while its user context is being
removed, the page is closed immediately.

* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
(WebKit::BidiBrowserAgent::BidiUserContextDeletionRecord::BidiUserContextDeletionRecord):
(WebKit::BidiBrowserAgent::didCreatePage):
(WebKit::BidiBrowserAgent::willClosePage):
(WebKit::BidiBrowserAgent::close):
(WebKit::BidiBrowserAgent::createUserContext):
(WebKit::BidiBrowserAgent::getUserContexts):
(WebKit::BidiBrowserAgent::removeUserContext):
* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.h:
* Source/WebKit/UIProcess/Automation/BidiUserContext.cpp:
(WebKit::BidiUserContext::allPages const):
* Source/WebKit/UIProcess/Automation/BidiUserContext.h:
* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::didCreatePage):
(WebKit::WebAutomationSession::willClosePage):
* Source/WebKit/UIProcess/Automation/WebAutomationSession.h:
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::m_pageForTesting):

Canonical link: https://commits.webkit.org/294938@main
